tests.md

```diff
@@ -0,0 +1,219 @@
+  u_cidr    = "10.0.${local.u_number}.0/24"
+}
+```
+
+This way you can replace a `true` known bool value with `local.u_true` unknown bool value with ease. (The `local.u_true`
+is always equal to `true`, because the `length()` is always greater than zero. But only the Apply stage determines
+that, the Plan stage assumes it can be either true or false.) Same for other types of inputs: you can replace any string
+by `local.u_string`, and number by `local.u_number`, and so on. This way it is possible to set every input of the module
+to an unknown value in quite a readable way.
+
+This pattern uses `random_pet`, a very simple Hashicorp-provided resource which simply generates names,
+e.g. "relevant-horse". It is very fast, does not communicate to any cloud, and does not incur any cost, but otherwise
+does not bring any real functionality here - the code above would be usable even if we would use `aws_vpc` instead.
+
+### Outputs
+
+The majority of module's outputs should be consumable by subsequent Terraform code. Only the minority of outputs
+exist to be merely displayed to a user when Terraform completes. We can test for some very predictable scenarios of how
+our module's output is consumed by some typical user-side code. It turns out that unknown values impact this too.
+
+The main scenario is when an output is a `map`. There is not much one can do with a map, except to perform a series
+of transformations and finally feed it to some `for_each` block at some point. But it's just impossible if the keys
+of the map are an unknown set - the Plan will just fail with "Invalid for_each argument". The module theoretically works
+but prevents any Lego®-style building. (Such a map output can still be displayed to users however.)
+
+```hcl2
+module "mymodule" {
+  source = "../../modules/mymodule"
+
+  /* inputs */
+}
+
+### Feed all the map outputs to a dummy for_each ###
+
+resource "random_pet" "consume_mymodule_my_map_output" {
+  for_each = module.mymodule.my_map_output
+}
+
+resource "random_pet" "consume_mymodule_second" {
+  for_each = module.mymodule.second
+}
+
+resource "random_pet" "consume_mymodule_third" {
+  for_each = module.mymodule.third
+}
+```
+
+On Terraform 0.13 and above condense that to a single block:
+
+```hcl2
+resource "random_pet" "consume_maps" {
+  for_each = merge(
+    module.mymodule.my_map_output,
+    module.mymodule.second,
+    module.mymodule.third,
+
+    module.mymodule_alt.my_map_output,
+    module.mymodule_alt.second,
+    module.mymodule_alt.third,
+  )
+}
+```
+
+The `_alt` above refers to an alternate call to the same module, with a different set of known inputs. For example,
+if the `vpc` module has the input `create_vpc` which is true by default, an alternate call would have it false.
+This would test Plan on multiple execution paths within the same code.
+
+Similarly, sometimes an output is a list, but a user can still be predicted to convert it to a map anyway. Lists
```

Comment:
Not sure if I understand this line - do you mean that the user can predict the list values and convert them in to a map?

Reply:
Let's say a module.mymodule.subnet_ids is a list of strings `["subnet1", "subnet2"]`. You cannot feed a list to a for_each, Terraform never accepts that. But on the other hand, sometimes you just know that some module's users will try to do something like this:

```
resource "aws_route_table_association" "this" {
  for_each = { for v in module.mymodule.subnet_names: v => v }

  subnet_id = each.value
}
```

Better test that list against that future, if you can predict it. (My example output would fail such test and would need to be fixed/restructured.)

Ok, how can I make this line of doc better then?

Reply:
"You cannot feed a list to a for_each, Terraform never accepts that." - yeaaah but you can always convert a list to a set for strings :) 

IMO changing line 179 to something like that and adding the explanation you just provided to me would be beneficial.
`Similarly, sometimes an output is a list, but a user can still insist it to convert it to a map anyway.`

Reply:
I've modified, please re-review that part.

I don't want to even mention `toset()`, because it doesn't solve the problem.

I cannot see it myself when only looking at the code, but it only proves how beneficial this type of testing is. It will just fail and stop me:

```txt
│ Error: Invalid for_each argument
│ 
│   on main.tf line 331, in resource "random_pet" "set":
│  331:   for_each = toset([local.u_string, "subnet2"])
│     ├────────────────
│     │ local.u_string is a string, known only after apply
│ 
│ The "for_each" value depends on resource attributes that cannot be
│ determined until apply, so Terraform cannot predict how many instances will
│ be created. To work around this, use the -target argument to first apply
│ only the resources that the for_each depends on.
╵}
```

Reply:
Please re-review @damianfedeczko 😄